v5: Add .fs-* utilities for font-size and rename font-weight/-style utilities

mdo opened the pull request:
- Adds new font-sizes Sass map
- Generates six new classes for setting only font-size
- Updates docs to mention this, including a scss-docs reference

It occurs to me this might need some `rfs` magic, but I'm unsure how to add that right now to the utilities file. Also, should these be responsive so folks can do `.fs-md-3 .fs-lg-4`? I'm thinking not given obvious file bloat.

Also makes me think `.font-weight-*` utilities should be shortened to `.fw-*`. `.font-style` is obviously trickier... could remap to `.text-*` or `.font-*`?

Fixes #25832.

https://deploy-preview-30571--twbs-bootstrap.netlify.app/docs/5.0/utilities/text/#font-size

mdo committed:
v5: Add .fs-* utilities for font-size

- Adds new font-sizes Sass map
- Generates six new classes for setting only font-size
- Updates docs to mention this, including a scss-docs reference

MartijnCuppens commented:
> It occurs to me this might need some `rfs` magic, but I'm unsure how to add that right now to the utilities file. 

Just add `rfs: true` to the utility map, and we should be fine.

> Also, should these be responsive so folks can do `.fs-md-3 .fs-lg-4`? I'm thinking not given obvious file bloat.

Since RFS already rescales the font sizes, I don't think there's any need for this.

> Also makes me think .font-weight-* utilities should be shortened to .fw-*. .font-style is obviously trickier... could remap to .text-* or .font-*?

Yup, `fw` seems appropriate, `text-*` might be confusing, especially for `font-weight: normal`.

---

I think it can be quite confusing that the margin of `m-2` is larger than `m-1`, but `fs-2` is smaller than `fs-1`. Also, if you want to add font sizes larger than `fs-1`, you'll need to add `fs-0`, which doesn't really make sense.

It might be better to work with `sm/md/lg` modifiers like this:

```scss
$font-sizes: (
  "xs": .75rem,
  "sm": .875rem,
  "md": 1rem,
  "lg": 1.25rem,
  "xl": 1.5rem,
  "xxl": 2rem
) !default;
```

mdo committed:
Update font utilities

- Make .fs-* utils for font-size use RFS
- Rename .font-weight-* utils to .fw-*
- Rename .font-style-* utils to .f-*
- Update order of utilities a bit
- Update docs to match

mdo committed:
Update migration docs to make note of these changes

MartijnCuppens commented on scss/_utilities.scss:
```diff
@@ -358,8 +358,19 @@ $utilities: map-merge(
+    ),
+    "font-style": (
+      property: font-style,
+      class: f,
```
I don't think `f-normal` is clear enough, maybe use the `fst` prefix (or something else?)

mdo replied:
Yeah, playing with this for sure still. Tailwind does `.italic`/`.no-italic`, the latter of which makes me laugh a little bit haha. Wanted to keep it somewhat familiar and `.fst` does fit nice enough. Any other ideas?

MartijnCuppens replied:
Not atm. `no-italic` feels very "unlike any other class".

mdo commented:
> It might be better to work with `sm/md/lg` modifiers like this:

I've been avoiding new classes that mix with our breakpoints notation to see if we can clear up "scales". I chose this approach since it maps 1:1 with our `font-size`s on headings. I know we did it with `line-height`, but maybe we change that? Hmm...

mdo committed:
Be more specific in font-size docs about scale

mdo committed:
Update font-style abbreviation to .fst

ffoodd commented on scss/_utilities.scss:
```diff
@@ -358,8 +358,19 @@ $utilities: map-merge(
+    ),
     "font-weight": (
-      property: font-weight,
+      property: fw,
```
Shouldn't `fw` be the class, and `property` stick to `font-weight`?

mdo committed:
Fix font-weight property

ffoodd commented:
Naming is pretty hard here, my two cents: 
1. if map keys are 1-6, why keep `$h*-font-size` variable at all: they could be replaced by a loop in our reboot.
2. if not, couldn't we handle a single map with all our font-sizes, from small to `display-1`—but I don't have any clue of an accurate naming for now -_-

We could stick with numbers (like spacers) but it wouldn't match headings hierarchy anymore: but should it?

Thinking out loud for now, but there's more to be done I guess.

mdo commented:
Going to hold on this for now with that last comment.

I'm wondering if we can build a single type scale system (GitHub/Primer does `h000`, `h00`, and `h0`) to go with all our type sizes and then nix the display classes. Unsure if it makes sense to change more classes or just add on.

mdo commented:
Decided against the above. Let's leave the display classes as-is. v6 can tackle a single set of font-sizes if we still want it.

Just rebased and resolved the conflicts here. This should be ready for review now!

ffoodd commented on site/content/docs/5.0/migration.md:
```diff
@@ -246,12 +246,19 @@ Badges were overhauled to better differentiate themselves from buttons and to be
+### Fonts and text
+- Added `.fs-*` utilities for `font-size`.
+- Renamed `.font-weight-*` utilities as `.fw-*` for brevity.
+- Renamed `.font-style-*` utilities as `.f-*` for brevity.
```
Your map sets `fst` as a class name for `font-style`.

MartijnCuppens replied:
Fixed that

MartijnCuppens committed:
Formatting & font style utility fix

XhmikosR commented:
@mdo this needs to move the migration section to alpha3 :)

mdo committed:
Move to Alpha 3 migration section

XhmikosR commented on site/content/docs/5.0/migration.md:
```diff
@@ -7,6 +7,17 @@ aliases: "/migration/"
+- Added `.fs-*` utilities for `font-size`.
+- Renamed `.font-weight-*` utilities as `.fw-*` for brevity.
+- Renamed `.font-style-*` utilities as `.fst-*` for brevity.
+- Renamed `.text-monospace` to `.font-monospace`.
+- New `line-height` utilities: `.lh-1`, `.lh-sm`, `.lh-base` and `.lh-lg`. See [here]({{< docsref "/utilities/text#line-height" >}}).
+- Removed `.text-hide` as it's an antiquated method for hiding text that shouldn't be used anymore.
```
I think these changes belong to alpha2?

XhmikosR commented on site/content/docs/5.0/utilities/text.md:
```diff
@@ -63,18 +63,35 @@ Transform text in components with text capitalization classes.
+<p class="fw-light">Light weight text.</p>
+<p class="fw-lighter">Lighter weight text (relative to the parent element).</p>
+<p class="fst-italic">Italic text.</p>
+<p class="fst-normal">Text without font style</p>
```
Maybe we should reword this? It does have a `font-style`, it's just set to `normal`.

stormitornado commented:
I propose to name new classes as shortcuts in emmet:
fz - font-size
fs - font-style
fw - font-weight

mdo committed:
Update migration guide for more details, splitting alpha 2 stuff back to the appropriate section in Migration guide

mdo committed:
Tweak language on example

MartijnCuppens commented on site/content/docs/5.0/migration.md:
```diff
@@ -21,10 +21,22 @@ toc: true
 
 - Introduce `$enable-smooth-scroll`, which applies `scroll-behavior: smooth` globally—except for users asking for reduced motion through `prefers-reduced-motion` media query. [See #31877](https://github.com/twbs/bootstrap/pull/31877)
 
+<<<<<<< HEAD
```
![Oh no](https://media.giphy.com/media/14aUO0Mf7dWDXW/giphy.gif)
